pygpu/collectives.pyx

```diff
@@ -0,0 +1,299 @@
+from libc.stdlib cimport malloc, calloc, free
+
+from pygpu.gpuarray cimport (gpucontext, GpuContext, _GpuArray, GpuArray,
+                             ensure_context,
+                             GA_NO_ERROR, get_exc, gpucontext_error,
+                             GpuArray_IS_C_CONTIGUOUS,
+                             GA_C_ORDER, GA_F_ORDER, GA_ANY_ORDER,
+                             pygpu_empty_like, pygpu_empty, memcpy)
+from pygpu.gpuarray import GpuArrayException
+
+
+cdef class GpuCommCliqueId:
+    """
+    """
+    def __cinit__(self, GpuContext context=None):
+        self.context = ensure_context(context)
+
+    property comm_id:
+        "Unique clique id to be used by each GpuComm in a group of devices"
+        def __get__(self):
+            cdef bytes res
+            comm_generate_id(self.context.ctx, self)
+            res = self.comm_id.internal[:GA_COMM_ID_BYTES]
+            return res
+
+        def __set__(self, unsigned char[:] bytearr):
+            cdef int length
+            length = bytearr.shape[0]
+            if length < GA_COMM_ID_BYTES:
+                raise ValueError, "GpuComm clique id must have length %d bytes" % (GA_COMM_ID_BYTES)
+            memcpy(self.comm_id.internal, <char*>&bytearr[0], GA_COMM_ID_BYTES)
+
+
+cdef class GpuComm:
+    """
+    """
+    def __dealloc__(self):
+        gpucomm_free(self.c)
+
+    def __reduce__(self):
+        raise RuntimeError, "Cannot pickle GpuComm object"
+
+    def __cinit__(self, GpuCommCliqueId cid not None, int ndev, int rank):
+        self.context = cid.context
+        cdef int err
+        err = gpucomm_new(&self.c, self.context.ctx, self.cid.comm_id,
+                          ndev, rank)
+        if err != GA_NO_ERROR:
+            raise get_exc(err), gpucontext_error(self.context.ctx, err)
+
+    def get_count(self):
+        cdef int gpucount
+        comm_get_count(self, &gpucount)
+        return gpucount
+
+    def get_rank(self):
+        cdef int gpurank
+        comm_get_rank(self, &gpurank)
+        return gpurank
+
+    def reduce(self, GpuArray src not None, op, GpuArray dest=None, int root=-1):
+        if dest is None:
+            if root != -1:
+                return comm_reduce_from(self, src, to_reduce_opcode(op), root)
+            else:
+                return pygpu_make_reduced(self, src, to_reduce_opcode(op))
+        if root == -1:
+            comm_get_rank(self, &root)
```

Comment:
Ok but since you get get the rank from the comm object you don't have to require that the user passes it in.  This is the sort of "nice interface" things that we usually do for python bindings.